### PR TITLE
[WebGPU][CMake] Support executable target wgslc

### PR DESCRIPTION
**Pull request by dneto0**
#### a62fb1f7043fe913255f7f1954d0564ef29ce6b9
<pre>
[WebGPU][CMake] Support executable target wgslc
<a href="https://bugs.webkit.org/show_bug.cgi?id=283111">https://bugs.webkit.org/show_bug.cgi?id=283111</a>

Reviewed by Tadeu Zagallo.

Add ENABLE_WEBGPU CMake toggle, defaulted to off

Also port it to Clang on Linux x86-64:
- Don&apos;t assume building on Apple platform:
  The notify framework is available only on Darwin.
- Don&apos;t assume __fp16 always exists. Use __Float16 instead.

Combined changes:
* Source/CMakeLists.txt: add WebGPU subdir if ENABLE_WEBGPU
* Source/WebGPU/CMakeLists.txt: Added.
* Source/WebGPU/WGSL/CMakeLists.txt: Added.
* Source/WebGPU/WGSL/ConstantValue.h:
  When __fp16 is absent, fall back to _Float16 on non-Apple platforms.
  _Float16 is in C++23, and commonly supported in Clang and GCC.
(WGSL::half::half):
(WGSL::ConstantValue::getHalf const):
* Source/WebGPU/WGSL/Metal/MetalCodeGenerator.cpp:
  Don&apos;t use Darwin&apos;s notify API on non-Apple targets.
(WGSL::Metal::dumpMetalCodeIfNeeded):
* Source/cmake/OptionsJSCOnly.cmake: Turn off ENABLE_WEBGPU
* Source/cmake/WebKitFeatures.cmake:
  Add ENABLE_WEBGPU toggle, defaulted off

Canonical link: <a href="https://commits.webkit.org/286825@main">https://commits.webkit.org/286825@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/de19c6d8cdeea683aad10f49c5c56aa8f28c8a1b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76989 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56024 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29904 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81537 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28268 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65172 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4320 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60336 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18414 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80056 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50280 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66084 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40648 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47682 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26594 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/70174 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68803 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23910 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82972 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/76267 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4369 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2931 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68610 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4524 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66057 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67861 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16978 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11848 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9928 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/98520 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4468 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21555 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4335 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7770 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6094 "Built successfully") | | |

### CODE REVIEW

File: Source/WebGPU/WGSL/ConstantValue.h [hunk 1]
```diff
@@ -36,22 +36,38 @@ namespace WGSL {
...
-    {
-    }
+        : value(val.value().getHalf().value) { }
 
     operator float() const
     {
         return static_cast<float>(value);
     }
 
-    __fp16 value { 0.f };
```
Comment (mwyrzykowski) [inline Source/WebGPU/WGSL/ConstantValue.h:54]: I think we are still compiling with C++20, can you use `__fp16` on `PLATFORM(COCOA)` and `_Float16` on Linux please?

Comment (mwyrzykowski) [reply Source/WebGPU/WGSL/ConstantValue.h:54]: or `PLATFORM(__APPLE__)` works too

File: Source/WebGPU/WGSL/ConstantValue.h [hunk 2]
```diff
@@ -36,22 +36,38 @@ namespace WGSL {
 #if HAVE(FP16_HALF_SUPPORT)
 using half = __fp16;
 #else
+// Wrap _Float16, which is the 16bit float type in C++23
```
Comment (mwyrzykowski) [inline Source/WebGPU/WGSL/ConstantValue.h]: I wonder if you could just use `<wtf/Float16.h>` instead, but for some reason it looks like `HAVE(FLOAT16)` is only 1 on `arm64`?

File: Source/WebGPU/WGSL/ConstantValue.h [hunk 3]
```diff
@@ -36,22 +36,38 @@ namespace WGSL {
...
 struct half {
+    using f16 = _Float16;
```
Comment (mwyrzykowski) [inline Source/WebGPU/WGSL/ConstantValue.h:46]: will using `_Float16` instead of `__fp16` give different results? Since `__fp16` gets promoted to float for operations, but I don't know if we rely on this.

@tadeuzagallo do you recall?

**Review by mwyrzykowski — COMMENTED** [review]
These changes seem fine (thank you for the contribution), I defer to @tadeuzagallo for approval once the builds are green

File: Source/WebGPU/WGSL/ConstantValue.h (hunk 3 above)
Comment (mwyrzykowski) [reply Source/WebGPU/WGSL/ConstantValue.h:46]: So it appears `_Float16` is promoted to `float` similar to `__fp16`. So I think it will be fine

File: Source/WebGPU/WGSL/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+# Build wgslc.
+# For simplicity, build it as a monolithic executable.
+
+# Target 'wgsl-types' generates TypeDeclarations.h in the build tree.
+add_custom_command(
+    OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/TypeDeclarations.h
+    COMMAND /usr/bin/env ruby
```
Comment (aperezdc) [inline Source/WebGPU/WGSL/CMakeLists.txt]: Here you would want to use the `${RUBY_EXECUTABLE}` variable, which gets defined by CMake to location of the Ruby interpreter to use; and will work on other systems like Windows, too:

```suggestion
    COMMAND ${RUBY_EXECUTABLE}
```

File: Source/WebGPU/WGSL/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        ${CMAKE_CURRENT_SOURCE_DIR}/generator/main.rb
+        ${CMAKE_CURRENT_SOURCE_DIR}/TypeDeclarations.rb
+    )
+add_custom_target(wgsl-types
+    ALL
+    DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/TypeDeclarations.h)
+
+
+# Configure exectuable 'wgslc'
+
```
Comment (aperezdc) [inline Source/WebGPU/WGSL/CMakeLists.txt]: I think the comment is redundant, as it is already clear what the intention is. Also IIRC one empty line would follow the WebKit CMake style better:

```suggestion

```

File: Source/WebGPU/WGSL/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+    ./AST/ASTStringDumper.cpp
+    ./AST/ASTUnaryExpression.cpp
+    ./AST/ASTVisitor.cpp
+
+    ./Metal/MetalCodeGenerator.cpp
+    ./Metal/MetalFunctionWriter.cpp
+    )
+
+set(wgslc_HEADERS
+    ./API.h
```
Comment (aperezdc) [inline Source/WebGPU/WGSL/CMakeLists.txt]: Is the `./` prefix in file names really needed here? In theory CMake should already do the right thing and make relative paths depend on the current source directory.

**Review by aperezdc — COMMENTED** [review]
I left a couple of comments about the CMake parts&mdash;nothing too important, though.

File: Source/WebGPU/WGSL/ConstantValue.h (hunk 3 above)
Comment (tadeuzagallo) [reply Source/WebGPU/WGSL/ConstantValue.h:46]: I remember we tried multiple different types for f16, but can't remember the details. I think the issue might have been when using `__half` instead. I can check locally that the wgslc tests still pass with the patch.

File: Source/WebGPU/WGSL/ConstantFunctions.h [hunk 7]
```diff
@@ -165,31 +165,31 @@ static ConstantResult constantUnaryOperation(const FixedVector<ConstantValue>& a
     return scalarOrVector([&](auto& arg) -> ConstantResult {
         if constexpr (constraint & Constraints::Bool) {
             if (auto* boolean = std::get_if<bool>(&arg))
-                return { { fn(*boolean) } };
+                return { ConstantValue { fn(*boolean) } };
```
Comment (tadeuzagallo) [inline Source/WebGPU/WGSL/ConstantFunctions.h]: The original code is a bit hacky, but the reason this won't work is that `fn(...)` might return a primitive value, a `ConstantValue` or a `ConstantResult`, so you can't always construct a `ConstantValue` from it.

File: Source/WebGPU/WGSL/ConstantValue.h (hunk 3 above)
Comment (tadeuzagallo) [reply Source/WebGPU/WGSL/ConstantValue.h:46]: I had to make some changes to get it compiling, but after that I ran all wgslc tests and it looks like it's fine.

**Review by mwyrzykowski — DISMISSED** [review]
marking request changes due to the `ConstantValue` / `ConstantResult` comment from @tadeuzagallo , agreed it is an issue and will break the error mechanisms I think

File: Source/WebGPU/WGSL/ConstantValue.h (hunk 2 above)
Comment (dneto0) [reply Source/WebGPU/WGSL/ConstantValue.h]: Right. HAVE_FP16_HALF_SUPPORT is defined if and only if CPU is `arm64`.

I didn't want to overly complicate the patch by doing a more granular check for the specific feature.

File: Source/WebGPU/WGSL/CMakeLists.txt (hunk 6 above)
Comment (dneto0) [reply Source/WebGPU/WGSL/CMakeLists.txt]: Heh, they aren't needed.  I basically piped a `find` call.  Will remove them.

File: Source/WebGPU/WGSL/ConstantFunctions.h (hunk 7 above)
Comment (dneto0) [reply Source/WebGPU/WGSL/ConstantFunctions.h]: I've tried it again. Without `ConstantValue` there GCC complains about ambiguous resolution of the ConstantResult (a.k.a. `WTF::Expected<ConstantValue,String>`) constructor.

Attaching 
[error.txt](https://github.com/user-attachments/files/17780762/error.txt)

but the gist of it is:
```
/project/webkit/WebKit/Source/WebGPU/WGSL/ConstantFunctions.h:168:43: error: conversion from ‘<brace-enclosed initializer list>’ to ‘WGSL::ConstantResult’ {aka ‘std::experimental::fundamentals_v3::expected<WGSL::ConstantValue, WTF::String>’} is ambiguous
  168 |                 return { { fn(*boolean) } };
      |                                           ^
In file included from /project/webkit/WebKit/build.release/WTF/Headers/wtf/text/StringImpl.h:35,
                 from /project/webkit/WebKit/build.release/WTF/Headers/wtf/text/WTFString.h:32,
                 from /project/webkit/WebKit/Source/WebGPU/WGSL/CompilationMessage.h:30,
                 from /project/webkit/WebKit/Source/WebGPU/WGSL/WGSL.h:28,
                 from /project/webkit/WebKit/Source/WebGPU/WGSL/TypeCheck.h:28,
                 from /project/webkit/WebKit/Source/WebGPU/WGSL/TypeCheck.cpp:27:
/project/webkit/WebKit/build.release/WTF/Headers/wtf/Expected.h:301:15: note: candidate: ‘constexpr std::experimental::fundamentals_v3::expected< <template-parameter-1-1>, <template-parameter-1-2> >::expected(unexpected_type&&) [with T = WGSL::ConstantValue; E = WTF::String; unexpected_type = std::experimental::fundamentals_v3::unexpected<WTF::String>]’
  301 |     constexpr expected(unexpected_type&& u) : base(__expected_detail::error_tag, std::forward<unexpected_type>(u).value()) { }
      |               ^~~~~~~~
/project/webkit/WebKit/build.release/WTF/Headers/wtf/Expected.h:300:15: note: candidate: ‘constexpr std::experimental::fundamentals_v3::expected< <template-parameter-1-1>, <template-parameter-1-2> >::expected(const unexpected_type&) [with T = WGSL::ConstantValue; E = WTF::String; unexpected_type = std::experimental::fundamentals_v3::unexpected<WTF::String>]’
  300 |     constexpr expected(const unexpected_type& u) : base(__expected_detail::error_tag, u.value()) { }
      |               ^~~~~~~~
/project/webkit/WebKit/build.release/WTF/Headers/wtf/Expected.h:297:15: note: candidate: ‘constexpr std::experimental::fundamentals_v3::expected< <template-parameter-1-1>, <template-parameter-1-2> >::expected(value_type&&) [with T = WGSL::ConstantValue; E = WTF::String; value_type = WGSL::ConstantValue]’
  297 |     constexpr expected(value_type&& e) : base(__expected_detail::value_tag, std::forward<value_type>(e)) { }
      |               ^~~~~~~~
/project/webkit/WebKit/build.release/WTF/Headers/wtf/Expected.h:296:15: note: candidate: ‘constexpr std::experimental::fundamentals_v3::expected< <template-parameter-1-1>, <template-parameter-1-2> >::expected(const value_type&) [with T = WGSL::ConstantValue; E = WTF::String; value_type = WGSL::ConstantValue]’
  296 |     constexpr expected(const value_type& e) : base(__expected_detail::value_tag, e) { }
```

I'm continuing to look but it might be something like this GCC bug. (but it's not clear if it's a bug in GCC).
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=59389#c8 

I'm not 100% clear on it but it might be a confusion about interpreting the inner pair of braces as constructing a ConstantResult and then the outer pair as invoking ConstantResult's ordinary copy or move constructor.

I'll try Clang.

Comment (dneto0) [reply Source/WebGPU/WGSL/ConstantFunctions.h]: FYI: I'm using GCC 13.2.0

Comment (dneto0) [reply Source/WebGPU/WGSL/ConstantFunctions.h]: Looks like Clang on Linux handles the original code just fine.  I'm happy about that, and will only claim Clang works.

File: Source/WebGPU/WGSL/CMakeLists.txt (hunk 4 above)
Comment (dneto0) [reply Source/WebGPU/WGSL/CMakeLists.txt]: Fixed this.

File: Source/WebGPU/WGSL/CMakeLists.txt (hunk 6 above)
Comment (dneto0) [reply Source/WebGPU/WGSL/CMakeLists.txt]: fixed this.

File: Source/WebGPU/WGSL/ConstantFunctions.h (hunk 7 above)
Comment (dneto0) [reply Source/WebGPU/WGSL/ConstantFunctions.h]: Removed these changes, and only claim Clang works.  I am using Clang 16.0.6

**Review by mwyrzykowski — COMMENTED** [review]
Looks good from my side, I will let @tadeuzagallo formally approve as it is WGSL

File: Source/WebGPU/WGSL/ConstantValue.h [hunk 8]
```diff
@@ -36,22 +36,38 @@ namespace WGSL {
...
+#endif
     half()
     {
     }
 
-    half(auto val)
-        : value(static_cast<float>(val))
-    {
-    }
+    // Constructor from an arithmetic type. Use a template here because the
```
Comment (tadeuzagallo) [inline Source/WebGPU/WGSL/ConstantValue.h:52]: Are these changes required? I was originally trying to avoid implicit conversions as it makes it ambiguous when constructing a `ConstantValue`.

File: Source/WebGPU/WGSL/ConstantValue.h [hunk 9]
```diff
@@ -150,6 +166,21 @@ struct ConstantValue : BaseValue {
             return *abstractInt;
         RELEASE_ASSERT_NOT_REACHED();
     }
+    half getHalf() const
```
Comment (tadeuzagallo) [inline Source/WebGPU/WGSL/ConstantValue.h:169]: Same here, we shouldn't be implicitly converting from integral values to floats, as that can only be done via explicit operations according to the language semantics.

File: Source/WebGPU/WGSL/ConstantValue.h [hunk 10]
```diff
@@ -150,6 +166,21 @@ struct ConstantValue : BaseValue {
...
+        if (auto* u32 = std::get_if<uint32_t>(this))
+            return *u32;
+        if (auto* abstractInt = std::get_if<int64_t>(this))
+            return *abstractInt;
+        if (auto* f32 = std::get_if<float>(this))
+            return *f32;
+        if (auto* f64 = std::get_if<double>(this))
+            return *f64;
+        if (auto* b = std::get_if<bool>(this))
+            return *b;
```
Comment (tadeuzagallo) [inline Source/WebGPU/WGSL/ConstantValue.h]: You'll probably need a `RELEASE_ASSERT_NOT_REACHED();` here to make the compiler happy

**Review by tadeuzagallo — COMMENTED** [review]
Overall looks good, I think the bots are just unhappy about the missing `RELEASE_ASSERT_NOT_REACHED`. I have a couple other suggestions, but just small nits.

File: Source/WebGPU/WGSL/ConstantValue.h (hunk 8 above)
Comment (dneto0) [reply Source/WebGPU/WGSL/ConstantValue.h:52]: My first attempt at the patch listed individual types like `double` `int` etc.  It worked on GCC x86-64 but failed compilation on the Mac platforms with Clang because I had misted `long long int`. So I switched it to a template based on arithmetic types.

You can see the first attempt on the left here: https://github.com/WebKit/WebKit/compare/e99f5a6caf0e382f565e1d53886e680e75a1159c..b5e47f1991101cb3eb15466d30ee57a0c0271046#diff-4daad8a03b2ec07ff1020c4a221a2ae9c81e47f11c46d407f8c76d56f486844f

File: Source/WebGPU/WGSL/ConstantValue.h (hunk 9 above)
Comment (dneto0) [reply Source/WebGPU/WGSL/ConstantValue.h:169]: oh, do you mean WGSL semantics?  Sure, I can remove the integer cases.

File: Source/WebGPU/WGSL/ConstantValue.h (hunk 8 above)
Comment (dneto0) [reply Source/WebGPU/WGSL/ConstantValue.h:52]: When I remove conversions from integral types I get errors like this:

```
In file included from /project/webkit/WebKit/Source/WebGPU/WGSL/Lexer.cpp:29:
/project/webkit/WebKit/Source/WebGPU/WGSL/ConstantValue.h:227:12: error: no matching constructor for initialization of 'std::optional<half>'
    return { value };
           ^~~~~~~~~
/project/webkit/WebKit/Source/WebGPU/WGSL/Lexer.cpp:997:31: note: in instantiation of function template specialization 'WGSL::convertFloat<WGSL::half, long>' requested here
            if (auto result = convertFloat<half>(value))
```

File: Source/WebGPU/WGSL/ConstantValue.h (hunk 9 above)
Comment (dneto0) [reply Source/WebGPU/WGSL/ConstantValue.h:169]: I've removed the integer and bool cases.

File: Source/WebGPU/WGSL/ConstantValue.h (hunk 10 above)
Comment (dneto0) [reply Source/WebGPU/WGSL/ConstantValue.h]: Done.

File: Source/WebGPU/WGSL/ConstantValue.h (hunk 8 above)
Comment (tadeuzagallo) [reply Source/WebGPU/WGSL/ConstantValue.h:52]: Hum, not sure about the other errors, looks like it's due to an unfortunate implicit cast. It's existing code though, and it's nice that the code is a bit more explicit now, so I'm happy to land as is and I can look into the casts later

**Review by tadeuzagallo — APPROVED** [review]

**Comment by dneto0** [comment]
The build failure seems unrelated
```
/Volumes/Data/worker/tvOS-18-Build-EWS/build/WebKitBuild/Release-appletvos/SDKAdditions/System/Library/Frameworks/MediaToolbox.framework/Headers/MTAudioProcessingTap.h:15:10: fatal error: 'AudioToolbox/AudioToolbox.h' file not found
   15 | #include <AudioToolbox/AudioToolbox.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by mwyrzykowski** [comment]
> The build failure seems unrelated
> 
> ```
> /Volumes/Data/worker/tvOS-18-Build-EWS/build/WebKitBuild/Release-appletvos/SDKAdditions/System/Library/Frameworks/MediaToolbox.framework/Headers/MTAudioProcessingTap.h:15:10: fatal error: 'AudioToolbox/AudioToolbox.h' file not found
>    15 | #include <AudioToolbox/AudioToolbox.h>
>       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ```

Indeed, would you like me to merge the PR @dneto0 ?

**Comment by dneto0** [comment]
> Indeed, would you like me to merge the PR @dneto0 ?

Yes please, if you can. I am unable.

**Comment by mwyrzykowski** [comment]
> > Indeed, would you like me to merge the PR @dneto0 ?
> 
> Yes please, if you can. I am unable.

No problem :) just wanted to make sure you didn't want to make any more changes 👍 

Thanks again for the PR!
